rbd/encryption.go

```diff
@@ -0,0 +1,116 @@
+// RBD_ENCRYPTION_ALGORITHM_AES128: AES 128bits
+// RBD_ENCRYPTION_ALGORITHM_AES256: AES 256bits
+const (
+	RBD_ENCRYPTION_ALGORITHM_AES128 = EncryptionAlgorithm(C.RBD_ENCRYPTION_ALGORITHM_AES128)
+	RBD_ENCRYPTION_ALGORITHM_AES256 = EncryptionAlgorithm(C.RBD_ENCRYPTION_ALGORITHM_AES256)
+)
+
+// EncryptionFormatOptsLUKS1 and EncryptionFormatOptsLUKS2 are identical
+// structures at the moment, just as they are in the librbd api.
+// The purpose behind creating different identical structures, is to facilitate
+// future modifications of one of the formats, while maintaining backwards
+// compatibility with the other.
+
+// EncryptionFormatOptsLUKS1: Options required for LUKS v1
+type EncryptionFormatOptsLUKS1 struct {
+	Alg            EncryptionAlgorithm
+	Passphrase     string
+	PassphraseSize uint
+}
+
+// EncryptionFormatOptsLUKS2: Options required for LUKS v2
+type EncryptionFormatOptsLUKS2 struct {
+	Alg            EncryptionAlgorithm
+	Passphrase     string
+	PassphraseSize uint
+}
+
+// EncryptionOptions interface is used to encapsulate the different encryption
+// formats options and enable converting them from go to C structures.
+type EncryptionOptions interface {
+	cephEncryptionOptions() (C.rbd_encryption_options_t, C.size_t)
+}
+
+func (opts EncryptionFormatOptsLUKS1) cephEncryptionOptions() (C.rbd_encryption_options_t, C.size_t) {
+	var cOptsSize C.size_t
+	var cOptsPtr C.rbd_encryption_options_t
+	var cOpts C.rbd_encryption_luks1_format_options_t
+	cOpts.alg = C.rbd_encryption_algorithm_t(opts.Alg)
+	cOpts.passphrase = C.CString(opts.Passphrase)
+	cOpts.passphrase_size = C.ulong(opts.PassphraseSize)
+	cOptsSize = C.size_t(unsafe.Sizeof(cOpts))
+	cOptsPtr = C.rbd_encryption_options_t(unsafe.Pointer(&cOpts))
+	return cOptsPtr, cOptsSize
+}
+
+func (opts EncryptionFormatOptsLUKS2) cephEncryptionOptions() (C.rbd_encryption_options_t, C.size_t) {
+	var cOptsSize C.size_t
+	var cOptsPtr C.rbd_encryption_options_t
+	var cOpts C.rbd_encryption_luks2_format_options_t
+	cOpts.alg = C.rbd_encryption_algorithm_t(opts.Alg)
+	cOpts.passphrase = C.CString(opts.Passphrase)
+	cOpts.passphrase_size = C.ulong(opts.PassphraseSize)
+	cOptsSize = C.size_t(unsafe.Sizeof(cOpts))
+	cOptsPtr = C.rbd_encryption_options_t(unsafe.Pointer(&cOpts))
+	return cOptsPtr, cOptsSize
+}
+
+// EncryptionFormat: Create an encryption format header
+//
+// Implements:
+//  int rbd_encryption_format(rbd_image_t image,
+//                            rbd_encryption_format_t format,
+//                            rbd_encryption_options_t opts,
+//                            size_t opts_size);
+//
+// To issue an IO against the image, you need to mount the image
+// with libvirt/qemu using the LUKS format, or make a call to
+// rbd_encryption_load().
+func (image *Image) EncryptionFormat(format EncryptionFormat, opts EncryptionOptions) error {
+	var cOptsSize C.size_t
+	var cOptsPtr C.rbd_encryption_options_t
+	cOptsPtr, cOptsSize = opts.cephEncryptionOptions()
```

Comment:
You might need a `defer C.free` for `cOptsPtr`? Hmm, the function does not return a pointer to the object, so maybe not? :thinking:

Reply:
I am no go expert, but I agree with @phlogistonjohn.